### PR TITLE
inductor: add lowering for aten.slice_scatter (copy_, out= on slice views)

### LINKED ISSUE #2147: Implement slice_scatter operator to support functionalization of copy_ on tensor views
### What feature would you like to be added?

Implement the `slice_scatter` operator.

This operator should functionally update a slice of a tensor and return a new tensor (**no in-place mutation**). It is required to support PyTorch-style functionalization patterns, particularly those generated from `copy_` applied to tensor views (e.g., slices, `split` outputs).

### Expected semantics

```python
def slice_scatter(input, src, dim=0, start=None, end=None, step=1):
    out = input.clone()
    out[..., start:end:step, ...] = src
    return out
```

### Why is this feature needed?

`slice_scatter` is required to correctly execute graphs produced by **functionalization** in modern PyTorch compilation pipelines (e.g., `torch.compile` / Inductor).

When `copy_` is applied to a **view**, it is transformed into a functional representation using `slice_scatter`.

### Example

```python
tmp = dst.split(1, dim=0)[1]
tmp.copy_(src)
```

is transformed into:

```python
dst_new = slice_scatter(dst, src, dim=0, start=1, end=2)
```

Implementing this operator enables correct handling of view mutations and improves compatibility with PyTorch-style graph transformations.

### Any alternative solution that you may want to share?

An alternative would be to directly support in-place `copy_` on views.

However, this requires complex alias and mutation handling and does not fit well with the functional graph model used by compilers.

Therefore, implementing `slice_scatter` is the preferred approach.

### Anything else we need to know?

* `slice_scatter` is part of a broader class of **view-inverse operators** (e.g., `select_scatter`, `diagonal_scatter`) used to represent mutations on views.
* Compilers may later **re-lower** this back into in-place ops (reinplacement), but a correct functional version is required first.
* This operator is essential for supporting `copy_` on slices and split outputs.

Parent issue:
- #1333

### PR DESCRIPTION
**Pull request by yoheiueda**
#### What type of PR is this?

- [x] feature

#### What this PR does:

Registers a Spyre Inductor lowering for `aten.slice_scatter.default`, so slice writes now compile on Spyre. This covers the source-level forms that functionalize to `aten.slice_scatter`, for example a functional `slice_scatter`:

```python
torch.slice_scatter(base, src, dim, start, end)
```

a `copy_` into a slice view:

```python
base.narrow(dim, start, end - start).copy_(src)
```

and an `out=` write into a slice view:

```python
torch.add(src, src, out=base.narrow(0, start, end - start))
```

For `out=`, only contiguous slices are supported: Torch Inductor graph-breaks on a non-contiguous `out=` tensor (`unimplemented("Attempted to call op with non-contiguous out= tensor")`), and a slice view is contiguous only when it targets the outermost dim (dim 0).

PyTorch's in-tree functional decomposition of `slice_scatter` cannot codegen on Spyre — it hits the `index_expr` wall (`Cannot resolve target for 'index_expr'`). This lowering handles it directly:

1. `clone` + `realize` the base so the slice folds into a realized-storage `ReinterpretView`.
2. Build a `SliceView` over the sliced region — its layout carries the start offset.
3. `mutate_to` that view from the source.

Only unit step is supported; a non-unit step raises `Unsupported` (it does not map to a single `SliceView`).

Scope: stick-aligned writes on the stick (last) dim and full-stick writes on a non-last dim. Offset / sub-stick writes into internal buffers are **not** yet handled — they need mutation-target relocation and/or output padding and are addressed in a follow-up PR (stacked on the restickify-padding work, #3110). On this branch those shapes are not raised (baseline `main` already errored on every `slice_scatter`); the follow-up makes them correct.

#### Which issue(s) this PR is related to:

Fixes #2147

#### Special notes for your reviewer:

This is the first of a stacked split:

- **This PR** — lowering only, aligned coverage.
- Follow-up — internal-buffer mutation-target relocation + sub-stick / offset writes + multi-write / cat, stacked on #3110.

New tests in `tests/inductor/test_inductor_ops.py`, all comparing the compiled Spyre result against CPU. They cover the four source-level write styles that functionalize to `aten.slice_scatter` — `internal` (`slice_scatter(x + 1.0, ...)`), `graph_input` (`slice_scatter(x, ...)`, placeholder copy-back), `view_copy` (`narrow().copy_()`), and `out_arg` (`torch.add(..., out=base.narrow(...))`) — across 1d/2d/3d/4d shapes:

- Functional / `copy_` styles: stick-aligned writes on the stick (last) dim (`stick_*`, `stick_off_*`) and full-stick writes on a non-last dim (`nonstick_*`).
- `out=` style: contiguous dim-0 writes only (see the limitation noted above); `dim0_*` / `dim0_off_*` exercise aligned and offset dim-0 slices.

`test_slice_scatter_step_raises` asserts a non-unit-step slice raises. The CI op-shard config (`test_inductor_ops_misc_shape_a_config.yaml`) gains an explicit `test_slice_scatter.*` entry.

#### Does this PR introduce a user-facing change?

Yes — `torch.slice_scatter` / `Tensor.slice_scatter` now compile on Spyre for stick-aligned writes instead of failing to codegen.

#### Additional note:

This PR is a prerequisite for an upcoming PR that implements sub-stick torch.cat.
- #1464

### CODE REVIEW

**Comment by yoheiueda** [comment]
@dgrove-oss @avery-blanchard Could you take a look at this PR?

This PR is one of the prerequisites for an upcoming PR implementing sub-stick `torch.cat` #1464

File: torch_spyre/_inductor/lowering.py [hunk 1]
```diff
@@ -1067,6 +1067,28 @@ def lower_overwrite(input, output, dims, offsets):
...
+        raise Unsupported(
+            f"slice_scatter with step={step} is not supported on Spyre "
+            f"(only unit step maps to a SliceView mutation)"
+        )
+
+    start = 0 if start is None else start
+    end = size[dim] if end is None else end
+
+    output = lowering.clone(self)
+    output.realize()
```
Comment (avery-blanchard) [inline torch_spyre/_inductor/lowering.py:1086]: If possible, could we add a decomposition for slice_scatter to avoid needing to do a realize for the clone operation here in the lowering?

Comment (yoheiueda) [reply torch_spyre/_inductor/lowering.py:1086]: @avery-blanchard Thanks for the suggestion!

Did you have a specific decomposition in mind?

I took a look at this, but ran into a couple of issues:

- The upstream `slice_scatter` decomposition doesn't currently work for Spyre because it ultimately decomposes to ops we don't support natively yet (`aten.arange` and `aten.index`): https://github.com/pytorch/pytorch/blob/v2.13.0/torch/_decomp/decompositions.py#L831-L871
- A custom decomposition doesn't seem viable either. Inductor decompositions need to be purely functional, and I wasn't able to find a formulation of `slice_scatter` that both preserves its semantics and avoids introducing additional unsupported ops. The natural implementation is effectively a slice followed by a mutation (`copy_`).

Given those constraints, the best approach I've found so far is:

1. `clone` the base tensor into the output
2. `mutate_to` the sliced region with `src`

This does result in two data movements (the full copy plus the slice update). Since the natural implementation of `slice_scatter` is a slice mutation (`copy_` into a view), I wasn't able to find a viable functional decomposition and therefore implemented it as a lowering instead.

Given that `slice_scatter` is a functional op, the output must preserve the base tensor's values outside the updated slice region, so some form of materializing those values into the output is still required.

If there's another decomposition strategy you're thinking of, I'd be happy to explore it.

**Review by avery-blanchard — APPROVED** [review]
looks good to me, thanks for the decomposition clarification!
